packages/legend-query/src/components/QueryBuilderValueSpecificationEditor.tsx

```diff
@@ -534,75 +599,601 @@ export const LatestDatePrimitiveInstanceValueEditor: React.FC = () => (
+            />
+          );
+        case QUERY_BUILDER_DATE_OPTION.ABSOLUTE_DATE:
+          return (
+            <AbsoluteDateValueSpecificationEditor
+              graph={graph}
+              valueSpecification={valueSpecification}
+              updateValueSpecification={updateValueSpecification}
+              setDateOptionValue={setDateOptionValue}
+            />
+          );
+        case QUERY_BUILDER_DATE_OPTION.ABSOLUTE_TIME:
+          return (
+            <AbsoluteTimeValueSpecificationEditor
+              graph={graph}
+              valueSpecification={valueSpecification}
+              updateValueSpecification={updateValueSpecification}
+              setDateOptionValue={setDateOptionValue}
+            />
+          );
+        case QUERY_BUILDER_DATE_OPTION.START_DAY_OF_DATE:
+          return (
+            <CustomStartDayOfDateValueSpecificationEditor
+              graph={graph}
+              customDateAdjustOptionValue={generateDateOption(
+                valueSpecification,
+              )}
+              updateValueSpecification={updateValueSpecification}
+              setDateOptionValue={setDateOptionValue}
+            />
+          );
+        case QUERY_BUILDER_DATE_OPTION.PREVIOUS_DAY_OF_WEEK:
+          return (
+            <CustomPreviousDayOfWeekValueSpecificationEditor
+              graph={graph}
+              customDateAdjustOptionValue={generateDateOption(
+                valueSpecification,
+              )}
+              updateValueSpecification={updateValueSpecification}
+              setDateOptionValue={setDateOptionValue}
+            />
+          );
+        default:
+          return <></>;
       }
     };
 
     return (
       <div className="query-builder-value-spec-editor__date">
-        {valueSpecification.genericType.value.rawType === strictDate && (
+        {typeCheckOption.expectedType === strictDate && (
           <DatePrimitiveInstanceValueEditor
-            valueSpecification={valueSpecification}
+            valueSpecification={valueSpecification as PrimitiveInstanceValue}
             className={className}
           />
         )}
-        {valueSpecification.genericType.value.rawType === dateTime && (
-          <DateTimePrimitiveInstanceValueEditor
-            valueSpecification={valueSpecification}
-            className={className}
-          />
-        )}
-        {valueSpecification.genericType.value.rawType === latestDate && (
+        {typeCheckOption.expectedType === latestDate && (
           <LatestDatePrimitiveInstanceValueEditor />
         )}
-        {expectedType === date && (
-          <div className="query-builder-value-spec-editor__dropdown">
-            <CustomSelectorInput
-              placeholder="Choose a type..."
-              options={typeOptions}
-              onChange={changeType}
-              value={selectedType}
-              darkMode={true}
-            />
+        {(typeCheckOption.expectedType === date ||
```

Comment:
Given the amount of custom logic to do, I think we should move this into its own file, I'm thinking of `QueryBuilderCustomDatePicker` and we can move all logic from `QueryBuilderDateTimeHelper` there.